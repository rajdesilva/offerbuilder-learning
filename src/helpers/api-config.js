import { init as initApm } from "@elastic/apm-rum";

export const baseURL = {
  API_URL_BASE: process.env.REACT_APP_BASE_API_URL,
  LOGIN_URL: process.env.REACT_APP_LOGIN_URL,
  LOGOUT_URL: process.env.REACT_APP_LOGOUT_URL,
  PROFILE_URL: process.env.REACT_APP_USER_PROFILE_URL,
  HISTORY_URL: process.env.REACT_APP_USER_HISTORY_URL,
};

export const apm = initApm({
  // Set required service name (allowed characters: a-z, A-Z, 0-9, -, _, and space)
  serviceName: process.env.REACT_APP_APM_SERVICE_NAME,
  serverUrl: process.env.REACT_APP_APM_SERVICE_URL,

  // Set service version (required for sourcemap feature)
  serviceVersion: process.env.REACT_APP_APM_SERVICE_VERSION,
});