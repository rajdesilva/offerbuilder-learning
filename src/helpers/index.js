export * from './api-config';
export * from './history';
