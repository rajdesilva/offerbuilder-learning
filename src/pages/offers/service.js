import { baseURL } from "../../helpers";
import {
  getParsedOffers,
  getSearchAndFilterImagesQueryUrl,
  getSearchAndFilterQueryURL,
  updateReduxWithOfferData,
} from "../../helpers/utility";
import { store } from "../../redux/store";
import {
  marketingActions,
  newOfferActions,
  offerListSearchAndFilterActions,
} from "./actions";
import { offerImageSearchAndFilterActions } from "./actions/offerImageSearchAndFilterActions";

export const validateOfferId = (offerId) => {
  return window
    .fetchWrapper(`${baseURL.API_URL_BASE}offers/${offerId}/exist`)
    .then((response) => {
      return response;
    })
    .catch((error) => {
      return error.toString();
    });
};

export const getCloudinarySignature = () => {
  return window
    .fetchWrapper(`${baseURL.API_URL_BASE}cloudinary/signature`)
    .then((response) => {
      return response;
    })
    .catch((error) => {
      return error.toString();
    });
};

export const getCityFromLatLong = (latLong) => {
  const url = `https://maps.googleapis.com/maps/api/geocode/json?latlng=${latLong.lat},${latLong.lng}&sensor=true&key=AIzaSyDydZJ7RZF1SAXwg4pZqnFFskYL8CW5RqA`;
  return fetch(url)
    .then((response) => response.json())
    .then((response) => {
      return response;
    })
    .catch((error) => {
      return error;
    });
};

export const getLanguages = () => {
  store.dispatch({
    type: marketingActions.FETCH_LANGUAGE_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(`${baseURL.API_URL_BASE}languages`)
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: marketingActions.FETCH_LANGUAGE_SUCCESS,
          payload: window.getValue(response, "data.languages"),
        });
      } else {
        store.dispatch({
          type: marketingActions.FETCH_LANGUAGE_FAILURE,
          payload: [],
        });
      }
    })
    .catch((error) => {
      store.dispatch({
        type: marketingActions.FETCH_LANGUAGE_FAILURE,
        payload: [],
      });
      return [];
    });
};

export const createNewOffer = (newOfferInfo) => {
  const url = `${baseURL.API_URL_BASE}offers`;
  store.dispatch({
    type: newOfferActions.CREATE_NEW_OFFER_LOADING,
    payload: true,
  });
  return window
    .fetchWrapper(url, {
      method: "POST",
      body: JSON.stringify(newOfferInfo),
    })
    .then((response) => response)
    .catch((error) => {
      return error;
    });
};

export const saveEditedOffer = (updatedOfferInfo) => {
  const url = `${baseURL.API_URL_BASE}offers/${updatedOfferInfo.id}`;
  return window
    .fetchWrapper(url, {
      method: "PATCH",
      body: JSON.stringify(updatedOfferInfo),
    })
    .then((response) => response)
    .catch((error) => {
      return error;
    });
};

export const getOfferDetails = (offerId) => {
  store.dispatch({
    type: offerListSearchAndFilterActions.OFFER_DETAILS_LOADING,
  });
  window
    .fetchWrapper(`${baseURL.API_URL_BASE}offers/${offerId}`)
    .then((response) => {
      if (response.success) {
        let offerData = window.getValue(response, "data.offer");
        offerData = getParsedOffers(offerData);
        updateReduxWithOfferData(offerData);
        store.dispatch({
          type: offerListSearchAndFilterActions.OFFER_DETAILS_SUCCESS,
        });
      } else {
        store.dispatch({
          type: offerListSearchAndFilterActions.OFFER_DETAILS_FAILURE,
        });
      }
    })
    .catch((error) => {
      console.log(error.toString());
      store.dispatch({
        type: offerListSearchAndFilterActions.OFFER_DETAILS_FAILURE,
      });
    });
};

export const deleteOffer = (offerId) => {
  return window
    .fetchWrapper(`${baseURL.API_URL_BASE}offers/${offerId}`, {
      method: "DELETE",
    })
    .then((response) => response);
};

export const updateOffer = (offerId, status) => {
  return window
    .fetchWrapper(`${baseURL.API_URL_BASE}offers/${offerId}/${status}`, {
      method: "PUT",
    })
    .then((response) => response);
};

export const searchAndFilterOffer = (offerType) => {
  store.dispatch({
    type: offerListSearchAndFilterActions.OFFER_LIST_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(
      `${baseURL.API_URL_BASE}offers/search/` +
        getSearchAndFilterQueryURL(offerType)
    )
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: offerListSearchAndFilterActions.OFFER_LIST_SUCCESS,
          payload: window.getValue(response, "data"),
        });
      } else {
        store.dispatch({
          type: offerListSearchAndFilterActions.OFFER_LIST_FAILURE,
          payload: [],
        });
      }
    })
    .catch((error) => {
      console.error("error =", error.toString());
      store.dispatch({
        type: offerListSearchAndFilterActions.OFFER_LIST_FAILURE,
        payload: [],
      });
    });
};

export const getImagesFromProperties = (propertyCode) => {
  return window
    .fetchWrapper(`${baseURL.API_URL_BASE}search?propertyCode=${propertyCode}`)
    .then((response) => response);
};

export const searchAndFilterImages = () => {
  store.dispatch({
    type: offerImageSearchAndFilterActions.OFFER_IMAGE_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(
      `${baseURL.API_URL_BASE}images/search` +
        getSearchAndFilterImagesQueryUrl()
    )
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: offerImageSearchAndFilterActions.OFFER_IMAGE_SUCCESS,
          payload: window.getValue(response, "data"),
        });
      } else {
        store.dispatch({
          type: offerImageSearchAndFilterActions.OFFER_IMAGE_FAILURE,
          payload: [],
        });
      }
    })
    .catch((error) => {
      console.error("error =", error.toString());
      store.dispatch({
        type: offerImageSearchAndFilterActions.OFFER_IMAGE_FAILURE,
        payload: [],
      });
    });
};

export const uploadImage = (file) => {
  let myHeaders = new Headers();

  myHeaders.append("Accept", "application/json");
  let formdata = new FormData();
  formdata.append("file", file);
  return window
    .fetchWrapper(`${baseURL.API_URL_BASE}images/upload`, {
      method: "POST",
      headers: myHeaders,
      body: formdata,
    })
    .then((response) => response);
};
