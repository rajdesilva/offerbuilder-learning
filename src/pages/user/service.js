import { baseURL } from "../../helpers";
import { userActions } from "./actions/userActions";
import { store } from "../../redux/store";

export const fetchUserInfo = () => {
  const url = `${baseURL.API_URL_BASE}user/`;
  store.dispatch({
    type: userActions.FETCH_USER_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(url)
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: userActions.FETCH_USER_SUCCESS,
          payload: window.getValue(response, "data"),
        });
      } else {
        store.dispatch({
          type: userActions.FETCH_USER_FAILURE,
          payload: false,
        });
      }
    })
    .catch(() => {
      store.dispatch({
        type: userActions.FETCH_USER_FAILURE,
        payload: false,
      });
    });
};

export const getUsersList = () => {
  store.dispatch({
    type: userActions.FETCH_USER_MANAGEMENT_LIST_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(
      `${baseURL.API_URL_BASE}users/`
    )
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: userActions.FETCH_USER_MANAGEMENT_LIST_SUCCESS,
          payload: window.getValue(response, "data"),
        });
      } else {
        store.dispatch({
          type: userActions.FETCH_USER_MANAGEMENT_LIST_FAILURE,
          payload: [],
        });
      }
    })
    .catch((error) => {
      store.dispatch({
        type: userActions.FETCH_USER_MANAGEMENT_LIST_FAILURE,
        payload: [],
      });
    });
};

export const addNewUser = (newUserInfo) => {
  const url = `${baseURL.API_URL_BASE}users/user`;
  return window
    .fetchWrapper(url, {
      method: "POST",
      body: JSON.stringify(newUserInfo),
    })
    .then((response) => response)
    .catch((error) => {
      return error;
    });
};

export const deleteUser = (email, userRole) => {
  const url = `${baseURL.API_URL_BASE}users/user/${email}?roleNames=${userRole}`;
  return window
    .fetchWrapper(url, {
      method: "DELETE",
    })
    .then((response) => response)
    .catch((error) => {
      return error;
    });
};
