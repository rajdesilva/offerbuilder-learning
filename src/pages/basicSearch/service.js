import { browseSupplyActions, supplySearchActions } from "./actions";
import { store } from "./../../redux/store";
import { globalActions } from "../../redux/globalActions";
import { baseURL } from "../../helpers";
import {
  createSearchQueryURL,
  getSearchParamsToFetchMarketPrice,
} from "../../helpers/utility";
import { offerListSearchAndFilterActions } from "../offers/actions";

export const searchSupply = (routePostSearch, isEditFlow) => {
  store.dispatch({
    type: browseSupplyActions.BROWSE_SUPPLY_SEARCH_LOADING,
    payload: true,
  });
  store.dispatch({
    type: supplySearchActions.SET_INCLUDE_MARKET_PRICE_FLAG,
    payload: false,
  });

  window
    .fetchWrapper(`${baseURL.API_URL_BASE}search/` + createSearchQueryURL(routePostSearch))
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: browseSupplyActions.BROWSE_SUPPLY_SEARCH_SUCCESS,
          payload: window.getValue(response, "data.properties"),
          totalMatch: window.getValue(response, "data.totalMatchedProperties"),
          pageOffset: window.getValue(response, "data.pageOffset"),
          pageSize: window.getValue(response, "data.pageSize"),
        });
      } else {
        store.dispatch({
          type: browseSupplyActions.BROWSE_SUPPLY_SEARCH_FAILURE,
          payload: window.getValue(response, "error.apierror.message"),
        });
      }
    })
    .catch((error) => {
      store.dispatch({
        type: browseSupplyActions.BROWSE_SUPPLY_SEARCH_FAILURE,
        payload: error.toString(),
      });
      store.dispatch({
        type: globalActions.SET_ERROR,
        payload: error.toString(),
      });
    });
};

export const advancedSearch = (
  params,
  searchType,
  page,
  pageSize,
  sortBy,
  order,
  groupType
  ) => {
    return window.fetchWrapper(`${baseURL.API_URL_BASE}global/properties?address=${params.address}&city=${params.city}&countryCode=${params.countryCode}&distance=${params.distance}&hotelName=${params.hotelName}&latitude=${params.latitude}&longitude=${params.longitude}&zip=${params.zip}&pageOffset=${page}&pageSize=${pageSize}&propertyGroupType=${groupType}&sortCriteria=${sortBy}&sortOrder=${order}&globalSearchType=${searchType}`);

    //return window.fetchWrapper(`https://offerbuilder-api-dev.kognitiv.com/global/properties?distance=10&globalSearchType=BASIC&latitude=47.2816&longitude=15.48809&propertyGroupType=ALL`);
}

export const globalSearchResult =(params)=>{
  
  window
    .fetchWrapper(`/global/properties?address=${params.address}`)
    .then((response) => {
        console.log(response)
    })
    .catch((error) => {
      store.dispatch({
        type: supplySearchActions.SEARCH_CHANNELS_FAILURE,
        payload: [],
      });
      return [];
    });
}

export const getChannels = () => {
  store.dispatch({
    type: supplySearchActions.SEARCH_CHANNELS_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(`${baseURL.API_URL_BASE}portals`)
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: supplySearchActions.SEARCH_CHANNELS_SUCCESS,
          payload: window.getValue(response, "data.result"),
        });
      } else {
        store.dispatch({
          type: supplySearchActions.SEARCH_CHANNELS_FAILURE,
          payload: [],
        });
      }
    })
    .catch((error) => {
      store.dispatch({
        type: supplySearchActions.SEARCH_CHANNELS_FAILURE,
        payload: [],
      });
      return [];
    });
};

export const getBrands = async () => {
  store.dispatch({
    type: supplySearchActions.SEARCH_BRANDS_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(`${baseURL.API_URL_BASE}sources`)
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: supplySearchActions.SEARCH_BRANDS_SUCCESS,
          payload: window.getValue(response, "data.brands"),
        });
        store.dispatch({
          type: offerListSearchAndFilterActions.UPDATE_OFFER_SELECTORS,
          payload: window.getValue(response, "data.brands"),
        });
      } else {
        store.dispatch({
          type: supplySearchActions.SEARCH_BRANDS_FAILURE,
          payload: [],
        });
      }
      return window.getValue(response, "data.brands");
    })
    .catch((error) => {
      store.dispatch({
        type: supplySearchActions.SEARCH_BRANDS_FAILURE,
        payload: [],
      });
      return [];
    });
};

export const getCurrencies = async () => {
  store.dispatch({
    type: supplySearchActions.FETCH_CURRENCIES_LOADING,
    payload: true,
  });
  window
    .fetchWrapper(`${baseURL.API_URL_BASE}currencies`)
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: supplySearchActions.FETCH_CURRENCIES_SUCCESS,
          payload: window.getValue(response, "data.currencies"),
        });
      } else {
        store.dispatch({
          type: supplySearchActions.FETCH_CURRENCIES_FAILURE,
          payload: [],
        });
      }
    })
    .catch((error) => {
      store.dispatch({
        type: supplySearchActions.FETCH_CURRENCIES_FAILURE,
        payload: [],
      });
      return [];
    });
};

export const getMarketPricesForListOfProperties = () => {
  store.dispatch({
    type: supplySearchActions.FETCH_MARKET_PRICE_LOADING,
    payload: true,
  });

  window
    .fetchWrapper(
      `${baseURL.API_URL_BASE}search/marketprice/` +
        getSearchParamsToFetchMarketPrice()
    )
    .then((response) => {
      if (response.success) {
        store.dispatch({
          type: supplySearchActions.FETCH_MARKET_PRICE_SUCCESS,
          payload: window.getValue(response, "data"),
        });
      } else {
        store.dispatch({
          type: supplySearchActions.FETCH_MARKET_PRICE_FAILURE,
          payload: [],
        });
      }
    })
    .catch((error) => {
      store.dispatch({
        type: supplySearchActions.FETCH_MARKET_PRICE_FAILURE,
        payload: [],
      });
      return [];
    });
};
