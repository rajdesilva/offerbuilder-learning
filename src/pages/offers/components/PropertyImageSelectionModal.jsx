import React, { useEffect, useState, Fragment } from "react";
import { Checkbox, Col, Image, List, Modal, Row, Skeleton } from "antd";
import { FormattedMessage } from "react-intl";
import defaultImg from "../../../assets/images/default-img.svg";
import { getImagesFromProperties } from "../service";
import { useRouteMatch } from "react-router-dom";
import styles from "./css/PropertyImageSelectionModal.module.less";
import { cloneDeep } from "lodash";
import PropertySelectionCarousel from "./PropertySelectionCarousel";
import PropertyImagesSelected from "./PropertyImagesSelected";
import PropertyImageListPagination from "./PropertyImageListPagination";

export default function PropertyImageSelectionModal(props) {
  const [displayCarousel, setDisplayCarousel] = useState({
    display: false,
    startIndex: 0,
  });

  const isOfferEditPage = useRouteMatch({
    path: "/offers/edit/:offerId?",
    strict: true,
    sensitive: true,
  });
  const [selectedImages, setSelectedImages] = useState(
    props.defaultSelectedImages || []
  );
  const [imageList, setImageList] = useState(
    isOfferEditPage
      ? []
      : window.getValue(props, "property.images").slice(0, 10) || []
  );
  const [remoteImages, setRemoteImages] = useState([]);
  const [loading, setLoading] = useState(false);
  const handleOk = () => {
    props.setImagesForProperty(selectedImages);
    props.hideModal();
  };

  const cancelModal = () => {
    props.hideModal();
  };

  const hideCarousel = () => {
    setDisplayCarousel({
      display: false,
      startIndex: 0,
    });
  };

  const getImageNameToDisplay = (image) => {
    const t = image.split("/");
    return t[t.length - 1];
  };

  useEffect(() => {
    if (isOfferEditPage) {
      setLoading(true);
      getImagesFromProperties(window.getValue(props, "property.propertyCode"))
        .then((response) => {
          if (window.getValue(response, "success")) {
            const images =
              window.getValue(
                response.data,
                "properties[0].mainProperty.info.images"
              ) || [];
            setRemoteImages(images);
            setImageList(images.slice(0, 10));
          } else {
            setImageList([]);
          }
          setLoading(false);
        })
        .catch((error) => {
          console.log("error = ", error.toString());
          setLoading(false);
          setImageList([]);
        });
    }
  }, []);

  const updateSelectedImages = (imageSelected, isToAdd) => {
    if (isToAdd) {
      const temp = cloneDeep(selectedImages);
      temp.push(imageSelected);

      setSelectedImages(temp);
    } else {
      const temp = selectedImages.filter((image) => image !== imageSelected);

      setSelectedImages(temp);
    }
  };

  return (
    <Modal
      title={
        <Fragment>
          <FormattedMessage id="nemo.selectingImagesFor" />
          &nbsp;{window.getValue(props, "property.hotelName")}
        </Fragment>
      }
      visible={true}
      closable={true}
      className={styles.propertyImageModal}
      width={1080}
      centered
      destroyOnClose={true}
      onOk={handleOk}
      okText={<FormattedMessage id="nemo.submit" />}
      onCancel={cancelModal}
      data-testid="image-selection-for-property-modal"
    >
      <div className={styles.modalContent}>
        {selectedImages && selectedImages.length > 0 ? (
          <PropertyImagesSelected
            setDisplayCarousel={setDisplayCarousel}
            setSelectedImages={setSelectedImages}
            selectedImages={selectedImages}
            setCarouselImage={props.setCarouselImage}
          />
        ) : (
          <Fragment />
        )}
        <div
          className={
            window.getValue(props, "imageList.length") === 0
              ? styles.galleryEmptyView
              : styles.galleryView
          }
        >
          <Skeleton loading={loading} active>
            <Checkbox.Group
              defaultValue={selectedImages}
              value={selectedImages}
            >
              <List
                grid={{ gutter: 16, column: 8 }}
                dataSource={imageList}
                loading={loading}
                renderItem={(image, index) => {
                  return (
                    <List.Item
                      key={index + "-" + image}
                      className={styles["marketing-image-hover"]}
                    >
                      <Image
                        src={image}
                        key={index + "--" + image}
                        onError={(ev) => (ev.target.src = defaultImg)}
                        preview={false}
                        className={styles.propertyImage}
                        onClick={() =>
                          setDisplayCarousel({
                            display: true,
                            startIndex: index,
                          })
                        }
                      ></Image>
                      <Checkbox
                        key={index + image}
                        className={styles.selectionCheckbox}
                        value={image}
                        onChange={(e) =>
                          updateSelectedImages(image, e.target.checked)
                        }
                        data-testid="language-setting-checkbox"
                      ></Checkbox>
                      <div className={styles.imgName}>
                        {getImageNameToDisplay(image) || (
                          <FormattedMessage id="nemo.unknown" />
                        )}
                      </div>
                    </List.Item>
                  );
                }}
              ></List>
            </Checkbox.Group>
          </Skeleton>
        </div>
        {displayCarousel.display ? (
          <PropertySelectionCarousel
            images={imageList}
            hideCarousel={hideCarousel}
            selectedImages={selectedImages}
            startIndex={displayCarousel.startIndex}
            updateSelectedImages={updateSelectedImages}
          />
        ) : (
          <Fragment />
        )}
      </div>
      <Col>
        <Row>
          <PropertyImageListPagination
            images={
              isOfferEditPage
                ? remoteImages
                : window.getValue(props, "property.images")
            }
            setImages={setImageList}
          />
        </Row>
      </Col>
    </Modal>
  );
}
