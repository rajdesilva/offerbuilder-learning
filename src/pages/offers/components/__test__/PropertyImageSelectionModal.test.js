import React from "react";
import { createStore } from "redux";
import { render, fireEvent, cleanup } from "../../../../helpers/testUtils";
import { MemoryRouter } from "react-router-dom";
import PropertyImageSelectionModal from "../PropertyImageSelectionModal";

import * as service from "../../service";
import { tempStore } from "./tempStore";

window.fetchWrapper = jest.fn().mockResolvedValue({
  success: true,
});

jest.mock("./../../service");

afterEach(cleanup);

describe("PropertyImageSelectionModal component test", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("create PropertyImageSelectionModal components snapshot", () => {
    const store = createStore(() => ({ ...tempStore }));
    store.dispatch = jest.fn();
    const props = {
      defaultSelectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      setImagesForProperty: () => {},
      hideModal: () => {},
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      propertyCode: "DEMO_JEEVAN",
    };

    const { container } = render(
      <MemoryRouter initialEntries={["/offers"]}>
        <div id="img-wrapper">
          <PropertyImageSelectionModal {...props} />
        </div>
      </MemoryRouter>,
      {
        store,
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("PropertyImageSelectionModal modal shows in offer edit flow, a submit button is clicked to hide the modal and set the state with images selected", () => {
    // Arrange
    const store = createStore(() => ({ ...tempStore }));
    store.dispatch = jest.fn();
    const props = {
      defaultSelectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      setImagesForProperty: jest.fn(),
      hideModal: jest.fn(),
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      propertyCode: "DEMO_JEEVAN",
    };

    // Act
    const { getByText } = render(
      <MemoryRouter initialEntries={["/offers/create-new-offer/4"]}>
        <div id="img-wrapper">
          <PropertyImageSelectionModal {...props} />
        </div>
      </MemoryRouter>,
      {
        store,
      }
    );
    // Assert
    expect(getByText("Selecting images for")).toBeTruthy();

    // Act
    fireEvent.click(getByText(/Submit/i));

    // Assert
    expect(props.hideModal).toHaveBeenCalledTimes(1);
    expect(props.setImagesForProperty).toHaveBeenCalledTimes(1);
  });

  test("PropertyImageSelectionModal modal shows in offer edit flow and a cancel button is clicked to hide the modal", () => {
    // Arrange
    const store = createStore(() => ({ ...tempStore }));
    store.dispatch = jest.fn();
    const props = {
      defaultSelectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      setImagesForProperty: jest.fn(),
      hideModal: jest.fn(),
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      propertyCode: "DEMO_JEEVAN",
    };
    // mocking get api when dialog is displayed
    service.getImagesFromProperties.mockImplementation((data) => {
      return Promise.resolve({
        success: true,
        data: {
          properties: [
            {
              propertyCode: "DEMO_JEEVAN",
              mainProperty: {
                info: {
                  name: "HOTEL DEMO_JEEVAN",
                  description:
                    "Experience award-winning service and sophisticated style at JW Marriott Hotel Vienna. Ideally situated between the airport and Vienna Expressway, our 5-star hotel is an ideal destination for both business travelers and vacationing families. Luxury rooms and suites showcase deluxe bedding and marble bathrooms, as well as high-speed internet, desks with ergonomic chairs and laptop-sized safes. Many accommodations also provide striking views of the Pune cityscape or the nearby mountain. Delightful dining options within the hotel include Italian fare, a vegetarian restaurant and a chic rooftop bar. Challenge yourself to a workout in the fitness center or a swim in the outdoor pool, followed by a treatment at the tranquil Quan Spa. If you're hosting an event in Pune, you can explore our 40,000 square feWe are always looking forward to children in our hotel. In our 120 sqm large playground your children will be supervised by professional trained staff. In the meantime the parents can just chill in our relax garden or in our sauna. Besides we provide a fitness room, an indoor swimming pool with panorama view and a cosmetic studio. In summer we offer guided excursions in the beautiful mountain world starting directly in front of our house and in winter you can explore the surrounding ski area with our guides.et of versatile, sophisticated venue space, including multiple boardrooms and a ballroom that can accommodate 2,000 people. We look forward to making your visit truly remarkable at JW Marriott Hotel Vienna.",
                  descriptions: [],
                  languages: ["DE", "EN"],
                  images: [
                    "https://res.cloudinary.com/seekda-dev/image/upload/w_1920,h_1080,c_fill,f_auto,fl_lossy,q_auto/development/DEMO_RAVI/seafacing.jpg",
                    "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
                  ],
                  city: "Wien",
                  countryName: "Austria",
                  trustyou: null,
                  award: { provider: "no-official-recognition", value: 4.0 },
                },
                propertyCode: "DEMO_JEEVAN",
                lcn: false,
                remainingCapital: null,
                supplier: null,
                channel: null,
                b2bPrice: null,
                b2cPrice: null,
                lowestMargin: null,
                highestMargin: null,
                arrivalDate: null,
                margins: [
                  {
                    b2bPrice: null,
                    b2cPrice: null,
                    marginOnPrice: null,
                    checkinDate: null,
                  },
                ],
              },
              alternateProperties: [],
            },
          ],
          pageSize: 100,
          pageOffset: 0,
          totalMatchedProperties: 1,
        },
        error: null,
      });
    });
    // Act
    const { getByText } = render(
      <MemoryRouter initialEntries={["/offers/edit/123"]}>
        <div id="img-wrapper">
          <PropertyImageSelectionModal {...props} />
        </div>
      </MemoryRouter>,
      {
        store,
      }
    );
    // Act
    fireEvent.click(getByText(/Cancel/i));

    // Assert
    expect(props.hideModal).toHaveBeenCalledTimes(1);
    expect(props.setImagesForProperty).toHaveBeenCalledTimes(0);
  });

  test("PropertyImageSelectionModal modal displayed in offer edit flow and  get images api is called only one time", () => {
    // Arrange
    const store = createStore(() => ({ ...tempStore }));
    store.dispatch = jest.fn();
    const props = {
      defaultSelectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      setImagesForProperty: jest.fn(),
      hideModal: jest.fn(),
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      propertyCode: "DEMO_JEEVAN",
    };
    // mocking get api when dialog is displayed
    service.getImagesFromProperties.mockImplementation((data) => {
      return Promise.resolve({
        success: true,
        data: {
          properties: [
            {
              propertyCode: "DEMO_JEEVAN",
              mainProperty: {
                info: {
                  name: "HOTEL DEMO_JEEVAN",
                  description:
                    "Experience award-winning service and sophisticated style at JW Marriott Hotel Vienna. Ideally situated between the airport and Vienna Expressway, our 5-star hotel is an ideal destination for both business travelers and vacationing families. Luxury rooms and suites showcase deluxe bedding and marble bathrooms, as well as high-speed internet, desks with ergonomic chairs and laptop-sized safes. Many accommodations also provide striking views of the Pune cityscape or the nearby mountain. Delightful dining options within the hotel include Italian fare, a vegetarian restaurant and a chic rooftop bar. Challenge yourself to a workout in the fitness center or a swim in the outdoor pool, followed by a treatment at the tranquil Quan Spa. If you're hosting an event in Pune, you can explore our 40,000 square feWe are always looking forward to children in our hotel. In our 120 sqm large playground your children will be supervised by professional trained staff. In the meantime the parents can just chill in our relax garden or in our sauna. Besides we provide a fitness room, an indoor swimming pool with panorama view and a cosmetic studio. In summer we offer guided excursions in the beautiful mountain world starting directly in front of our house and in winter you can explore the surrounding ski area with our guides.et of versatile, sophisticated venue space, including multiple boardrooms and a ballroom that can accommodate 2,000 people. We look forward to making your visit truly remarkable at JW Marriott Hotel Vienna.",
                  descriptions: [],
                  languages: ["DE", "EN"],
                  images: [
                    "https://res.cloudinary.com/seekda-dev/image/upload/w_1920,h_1080,c_fill,f_auto,fl_lossy,q_auto/development/DEMO_RAVI/seafacing.jpg",
                    "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
                  ],
                  city: "Wien",
                  countryName: "Austria",
                  trustyou: null,
                  award: { provider: "no-official-recognition", value: 4.0 },
                },
                propertyCode: "DEMO_JEEVAN",
                lcn: false,
                remainingCapital: null,
                supplier: null,
                channel: null,
                b2bPrice: null,
                b2cPrice: null,
                lowestMargin: null,
                highestMargin: null,
                arrivalDate: null,
                margins: [
                  {
                    b2bPrice: null,
                    b2cPrice: null,
                    marginOnPrice: null,
                    checkinDate: null,
                  },
                ],
              },
              alternateProperties: [],
            },
          ],
          pageSize: 100,
          pageOffset: 0,
          totalMatchedProperties: 1,
        },
        error: null,
      });
    });

    render(
      <MemoryRouter initialEntries={["/offers/edit/123"]}>
        <div id="img-wrapper">
          <PropertyImageSelectionModal {...props} />
        </div>
      </MemoryRouter>,
      {
        store,
      }
    );

    // Assert
    expect(service.getImagesFromProperties).toHaveBeenCalledTimes(1);
  });

  test("PropertyImageSelectionModal modal displayed and get images api is not called in offer creation flow", () => {
    // Arrange
    const store = createStore(() => ({ ...tempStore }));
    store.dispatch = jest.fn();
    const props = {
      defaultSelectedImages: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      setImagesForProperty: () => {},
      hideModal: jest.fn(),
      images: [
        "https://res.cloudinary.com/seekda-dev/image/upload/v1603359435/offer_builder_dev/gapj1ko3ffeuyhspdmbw.png",
      ],
      propertyCode: "DEMO_JEEVAN",
    };

    render(
      <MemoryRouter initialEntries={["/offers/create-new-offer/4"]}>
        <div id="img-wrapper">
          <PropertyImageSelectionModal {...props} />
        </div>
      </MemoryRouter>,
      {
        store,
      }
    );

    // Assert
    expect(service.getImagesFromProperties).toHaveBeenCalledTimes(0);
  });
});
