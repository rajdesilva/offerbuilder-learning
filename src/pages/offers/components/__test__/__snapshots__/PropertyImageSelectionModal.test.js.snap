// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyImageSelectionModal component test create PropertyImageSelectionModal components snapshot 1`] = `
<div
  id="img-wrapper"
/>
`;
